l1-contracts/contracts/bridgehub/Bridgehub.sol

```diff
@@ -0,0 +1,313 @@
+        return
+            IZkSyncStateTransition(stateTransition).l2TransactionBaseCost(
+                _gasPrice,
+                _l2GasLimit,
+                _l2GasPerPubdataByteLimit
+            );
+    }
+
+    /// @notice the mailbox is called directly after the baseTokenBridge received the deposit
+    /// @notice this assumes that either ether is the base token or
+    /// @notice the msg.sender has approved mintValue allowance for the baseTokenBridge.
+    /// @notice This means this is not ideal for contract calls, as the contract would have to handle token allowance.
+    function requestL2Transaction(
+        L2TransactionRequestDirect calldata _request
+    ) public payable override returns (bytes32 canonicalTxHash) {
+        {
+            address token = baseToken[_request.chainId];
+            // address tokenBridge = baseTokenBridge[_request.chainId];
+
+            if (token == ETH_TOKEN_ADDRESS) {
+                require(msg.value == _request.mintValue, "Bridgehub: msg.value mismatch");
+                // kl todo it would be nice here to be able to deposit weth instead of eth
+                IL1Bridge(baseTokenBridge[_request.chainId]).bridgehubDepositBaseToken{value: _request.mintValue}(
+                    _request.chainId,
+                    msg.sender,
+                    token,
+                    _request.mintValue
+                );
+            } else {
+                require(msg.value == 0, "Bridgehub: non-eth bridge with msg.value");
+                // note we have to pass token, as a bridge might have multiple tokens.
+                IL1Bridge(baseTokenBridge[_request.chainId]).bridgehubDepositBaseToken(
+                    _request.chainId,
+                    msg.sender,
+                    token,
+                    _request.mintValue
+                );
+            }
+        }
+
+        address stateTransition = getZkSyncStateTransition(_request.chainId);
+        canonicalTxHash = IZkSyncStateTransition(stateTransition).bridgehubRequestL2Transaction(
+            msg.sender,
+            _request.l2Contract,
+            _request.mintValue,
+            _request.l2Value,
+            _request.l2Calldata,
+            _request.l2GasLimit,
+            _request.l2GasPerPubdataByteLimit,
+            _request.factoryDeps,
+            _request.refundRecipient
+        );
+    }
+
+    /// @notice After depositing funds to the baseTokenBridge, the secondBridge is called
+    /// @notice to return the actual L2 message which is sent to the Mailbox.
+    /// @notice this assumes that either ether is the base token or
+    /// @notice the msg.sender has approved the baseTokenBridge with the mintValue,
+    /// @notice and also the necessary approvals are given for the second bridge.
+    /// @notice This function is great for contract calls to L2, the secondBridge can be any contract.
+    function requestL2TransactionTwoBridges(
```

Comment:
What would happen if the user deposits the chain with ether base token as the first and the second bridge?

Reply:
I think user will just lose the funds, or I am missing something?

Reply:
True, so bridgehubDeposit should not be called for baseToken

Reply:
So the first bridge sets mintValue, and second bridge only sets l2Contract and l2Calldata